openml/OpenML
# Minio upload api

#### Reference Issue



#### What does this PR implement/fix? Explain your changes.
- Modify the data upload API to support upload of parquet files. 
- Modify the data upload API to perform conversions using python scripts and upload both ARFF and parquet files to minio server.
-  The existing functionality of ARFF upload should not be affected.( except python API no other client API's dowload form minio yet)

**Implementation:**

**Upload arguments:**
- In the POST request,  ARFF file can be provided as usual using the field _**dataset**_ 
- parquet file can be provided using the new field _**dataset_pq**_
- We allow upload of either pq or ARFF (not both).

**CASE 1: ONLY ARFF is provided**
-  When ARFF is uploaded, we store it in the DATA folder(as always)
- After dataset id is generated, we convert ARFF to pq using a python script and then upload both pq and ARFF to minio server.
- Conversion and upload done in the same script`minio_upload_arff_to_pq.py`
Corresponding lines of code: https://github.com/openml/OpenML/blob/0903a01722a553682f81958a38f29cc9d7d9ad19/openml_OS/models/api/v1/Api_data.php#L1259

**CASE 2: Only pq is provided** 
- When pq is provided,  we convert it to ARFF  by invoking ` minio_convert_pq_to_arff.py`  (https://github.com/openml/OpenML/blob/0903a01722a553682f81958a38f29cc9d7d9ad19/openml_OS/models/api/v1/Api_data.php#L1102)
- We perform ARFF checks as usual and store the converted ARFF locally (in the data folder).  After this, dataset id is generated as usual. (https://github.com/openml/OpenML/blob/0903a01722a553682f81958a38f29cc9d7d9ad19/openml_OS/models/api/v1/Api_data.php#L1173)
- then upload both pq and ARFF to minio using  `minio_upload.py`. (https://github.com/openml/OpenML/blob/0903a01722a553682f81958a38f29cc9d7d9ad19/openml_OS/models/api/v1/Api_data.php#L1251)


#### How should this PR be tested?
ARFF as usual using the field **dataset**
pq is provided using the field **dataset_pq**
create POST requests and check minio server.

#### Any other comments?
We need to wait to merge this, we need a test minio server.
Apologies for numerous small commits, had to do this in order to test on the test server bit by bit
